### PR TITLE
Normalize symbol names

### PR DESCRIPTION
**Pull request by martint**
## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/planner/SymbolAllocator.java [hunk 1]
```diff
@@ -71,8 +71,7 @@ public Symbol newSymbol(String nameHint, Type type, @Nullable String suffix)
         requireNonNull(nameHint, "nameHint is null");
         requireNonNull(type, "type is null");
 
-        // TODO: workaround for the fact that QualifiedName lowercases parts
-        nameHint = nameHint.toLowerCase(ENGLISH);
+        nameHint = nameHint.toLowerCase(ENGLISH).replaceAll("[^a-z0-9$]+", "_");
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/SymbolAllocator.java:74]: - we should also remove repeated `_` characters (if present)
- not sure about regexp overhead. CharMatcher is more verbose, but might be faster, and we are here frequently

**Review by findepi — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/SymbolAllocator.java (hunk 1 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/planner/SymbolAllocator.java:74]: I'll look into that.

File: core/trino-main/src/main/java/io/trino/sql/planner/SymbolAllocator.java [hunk 2]
```diff
@@ -88,7 +93,7 @@ public Symbol newSymbol(String nameHint, Type type, @Nullable String suffix)
         String unique = nameHint;
 
         if (suffix != null) {
-            unique = unique + "$" + suffix;
+            unique = unique + "_" + suffix;
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/SymbolAllocator.java:96]: `$` was readable

but if you want to replace with `_`, do you need to update the `nameHint.lastIndexOf("_")` logic and the likes?

**Review by findepi — COMMENTED** [review]
i still think Symbol::Symbol should do checks as well
otherwise `ir.ExpressionFormatter` won't be able to assume symbol names do not contain unwanted characters. 

i.e. this is about class design. if `Symbol` class doesn't provide guarantees, `ir.ExpressionFormatter` should avoid relying on these guarantees.

File: core/trino-main/src/main/java/io/trino/sql/planner/SymbolAllocator.java (hunk 2 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/planner/SymbolAllocator.java]: Updated again with a few more changes that should address this.

**Comment by martint** [comment]
> i still think Symbol::Symbol should do checks as well
otherwise ir.ExpressionFormatter won't be able to assume symbol names do not contain unwanted characters.

Symbol doesn't care what characters are used. If the ExpressionFormatter has constraints, it can/should escape any characters it doesn't know how to render unambiguously.

**Comment by findepi** [comment]
> Symbol doesn't care what characters are used. If the ExpressionFormatter has constraints, it can/should escape any characters it doesn't know how to render unambiguously.

https://github.com/trinodb/trino/pull/21223

**Comment by findepi** [comment]
> Symbol doesn't care what characters are used. If the ExpressionFormatter has constraints, it can/should escape any characters it doesn't know how to render unambiguously.

https://github.com/trinodb/trino/pull/21223

**Review by dain — APPROVED** [review]

**Pull request merged**